### PR TITLE
New binhack/expression features

### PR DESCRIPTION
**Pull request by zero318**
So far this does four things:
1. Adds a new syntax to binhacks for creating constant pools, primarily to simplify using readonly data from codecaves.
2. Speeds up expression parsing by using a custom strtoul implementation that doesn't rely on errno/locale jank.
3. Adds more CPUID feature detection strings
4. Adds `@` as a way of referencing the current address from an expression (was previously possible with `-[4]`, but that isn't intuitive)

Ideally the locale dependency of `consume_float_value` could also be removed, but I'm not sure how to do that yet.

The `memalloc_low` file might seem a bit odd, but currently there's no way of making standard thcrap hackpoints work on x64, particularly since breakpoints must preserve all registers. There aren't any `CALL`/`JMP` opcodes with a 64 bit immediate offset and there's no reliable way of ensuring memory is allocated within the +/-2GB range of a 32 bit offset since DLLs could get loaded into the necessary range before we can use it. The only good solution is to allocate memory for pointers in the bottom 2GB of the address space and then use indirect calls/jumps with the abs32 SIB encoding that wasn't changed to be RIP relative. Since that involves creating a constant pool, I figured I would go ahead and preemptively include the necessary code.

### CODE REVIEW

**Comment by DankRank** [comment]
For floats, try std::from_chars from <charconv>. It's supposed to be locale independent. It can also convert ints, so, if you want, you can benchmark against that.

**Comment by zero318** [comment]
Unfortunately `std::from_chars` for integers doesn't have builtin support our `Rx`/`0b` notations and [the library implementation](https://github.com/microsoft/STL/blob/8f67ece4a654625220642c92fa029a5cfc77afa8/stl/inc/charconv#L235) is using essentially the same concept as my code, so the extra code necessary to support it seems to be negating any benefits.

It does seem promising for floats, but I haven't gotten to do much testing yet.

File: thcrap/src/ntdll.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+static void resolve_nt_funcs() {
+	const HMODULE ntdll_module = (HMODULE)GetModuleHandleA("ntdll.dll");
+	NtQueryInformationProcess = (NtQueryInformationProcessPtr)GetProcAddress(ntdll_module, "NtQueryInformationProcess");
+#if TH_X64
+	NtAllocateVirtualMemory = (NtAllocateVirtualMemoryPtr)GetProcAddress(ntdll_module, "NtAllocateVirtualMemory");
+	NtFreeVirtualMemory = (NtFreeVirtualMemoryPtr)GetProcAddress(ntdll_module, "NtFreeVirtualMemory");
+	RtlNtStatusToDosError = (RtlNtStatusToDosErrorPtr)GetProcAddress(ntdll_module, "RtlNtStatusToDosError");
+#endif
+}
```
Comment (32th-System) [inline thcrap/src/ntdll.cpp:47]: We could import those functions from ntdll.dll directly. The `lib` command from the Visual Studio command line tools can generate a .lib file for importing functions from DLLs directly with just a .def file. An `ntdll_internal_[x86/x64].def` file could simply be part of our source tree, with the command to turn it into a .lib file being a PreBuildEvent

I'm mainly putting this here for other developers. We did talk about this already but you mentioned that importing internal APIs from ntdll directly could increase the amount of AV detections.

File: thcrap/src/ntdll.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/**
```
Comment (32th-System) [inline thcrap/src/ntdll.h:1]: With this file now being present, we could also move the internal Windows structs (which right now are defined in util.h) into this header as well

File: thcrap/src/str_to_addr.asm [hunk 3]
```diff
@@ -0,0 +1,137 @@
+/**
```
Comment (32th-System) [inline thcrap/src/str_to_addr.asm:1]: For the other devs: this file was (in part) generated using the Clang compiler. Considering MSVC's problems, perhaps we could have a deeper discussion about switching the whole project over to Clang?

**Review by brliron — COMMENTED** [review]
I finally got enough time to look at this PR. Now, let's look at the 41 other unread messages in #github on Discord...

File: thcrap/src/str_to_addr.asm [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	.intel_syntax
+	.global @str_to_addr_impl@8
+
+/*
+; Fastcall
+; ECX: Base
+; EDX: str
+*/
+.align 16, 0xCC
+@str_to_addr_impl@8:
```
[the remark this answers is not in the document]
Comment (zero318) [reply thcrap/src/str_to_addr.asm:106]: Comment added

**Review by brliron — APPROVED** [review]

**Review by 32th-System — APPROVED** [review]

**Pull request merged**